Cmake setup to install at correct location

jeffin143 opened the pull request:
@birm 

```
install(DIRECTORY ${CMAKE_SOURCE_DIR}/include/
        DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/include)
```

This is not installing recursively , I mean only proto folder is getting installed and other files are not getting installed , Not sure what is the issue

The .hpp files are just mock files nothing else

jeffin143 committed:
initial cmake setup

birm commented on CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+
+ADD_SUBDIRECTORY(include/proto)
```
You've added proto as a subdir, but not filewriter. How do you intend to use/organise filewriter; is it just part of mlboard, or do you intend for it to be a separate library?

jeffin143 replied:
I intend it to be part of mlboard , every header file gets copied to include directory along with proto.pb.h and then user csn include the file <mlboard.hpp> 
Which will in turn have the file writer and everything imported

jeffin143 replied:
Just like ensmallen.hpp which imports multiple files of ensmallen_bits folder

birm replied:
Ok, can you explain more about your issue regarding installation; what do you want to happen, and what exactly is happening instead?

jeffin143 replied:
I wanted it to copy all my file in include/ so that any one can use #include<mlboard.hpp> and then can run the file

As of now it does copy the proto folder in include/ but all other folder gets ignored

birm commented on CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+  $<INSTALL_INTERFACE:include>)
+
+install(DIRECTORY ${CMAKE_SOURCE_DIR}/include/
+        DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/include)
```
```suggestion
        DESTINATION ${CMAKE_INSTALL_PREFIX}/include)
```

birm reviewed (approved):
This fixes it for me. Let me know if it works for you as well.

birm reviewed (changes requested):
Misclick earlier...

jeffin143 committed:
Add Scaler Summary

toshal-a commented on include/filewriter/crc_impl.hpp:
```diff
@@ -0,0 +1,242 @@
+// http://stackoverflow.com/a/26612761
+static uint32_t crc_32_tab[] = { /* CRC polynomial 0xedb88320 */
+    /*
+0x00000000, 0x77073096, 0xee0e612c, 0x990951ba, 0x076dc419, 0x706af48f,
```
If I am not wrong this is quite similar to the code written below. I think it's better to not add it in comments.

jeffin143 replied:
Done

jeffin143 committed:
add docs and updated readme file

jeffin143 committed:
rename class name and add docs

jeffin143 committed:
rename class name and add docs

toshal-a commented on README.md:
```diff
@@ -1,2 +1,139 @@
+
+Following is a snippet that would log scaler for 4 steps in temp directory.
+
+You can compile the following snippet of code using : `g++ main.cpp /build/proto/libproto.a -lprotobuf -lpthread` 
```
Can we remove the compulsion to link libproto.a? Did you tried the link below link I sent you before? It might help over here.
https://stackoverflow.com/questions/41430991/cmake-with-protobuf-file-in-subdirectory

jeffin143 replied:
```
install(FILES ${CMAKE_CURRENT_BINARY_DIR}/include/proto/libproto.a
        DESTINATION ${CMAKE_INSTALL_PREFIX}/lib/)
```

 I tried this but still manual linking is needed, I did try it out yesterday too for this

jeffin143 replied:
I have added few lines in cmake at the end please take a look

jeffin143 replied:
Please hit cmake ../ and make install

and try compiling your file using 

```
 g++ main.cpp -lproto -lprotobuf -lpthread --std=c++11 -v 
```

toshal-a replied:
Okay, this looks good.  :+1: 
Just one more thing. It looks like `mlboard` will be dependent on protobuf library. Currently we are just checking whether the package is present or not. So I think we should mention that before install `mlboard` user should also install protobuf on it's own. And also link to proper installation of` protobuf` would be great.

birm committed:
add style check codes

birm commented:
I've added style checks in c4f3a028ea1bcb6ee2940c10f4db78cfa4d2a35a, but they aren't triggered automatically. This may be useful for you. If not, feel free to drop the commit or otherwise modify.

zoq commented:
We can run the same style checks as done for the other repositories, if you think that is helpful?

birm commented:
That would work. We may need to add include/proto as a ignore directory.

zoq commented:
Right, if someone likes to open a PR for this here: https://github.com/mlpack/jenkins-conf/blob/master/linter/lint.sh

zoq commented:
@mlpack-jenkins test this please

zoq commented:
Alright, let's see if that works.

zoq commented:
@mlpack-jenkins test this please

zoq commented:
@mlpack-jenkins test this please

birm commented:
I had modified the script in ./ci to display to stdout rather than make a report; probably better to use the existing setup in the jenkins conf repo.

birm committed:
re-add jenkins reports

birm commented:
There we go! Thanks @zoq !!

toshal-a commented:
Great, thanks @birm and @zoq.

jeffin143 commented:
@mlpack-jenkins test this please

jeffin143 commented:
@walragatver @birm Ready to be reviewed

jeffin143 commented:
@brim Sorry, I force pushed the branch without pulling, can you once again add the commit if possible ?

jeffin143 committed:
add cpp lint

jeffin143 committed:
make it executable

birm committed:
resolve merge conflict

birm committed:
Merge branch 'cmake-setup' of https://github.com/jeffin143/mlboard into cmake-setup

birm commented:
That merge look ok to you, @jeffin143 ?

toshal-a commented:
@jeffin143, I think it's better that we merge the linting code in a seperate PR. To avoid incident that happened yesterday. Let me know what you think.

toshal-a reviewed (commented):
I was just going through the repo. I have left some comments

jeffin143 committed:
correct styles

jeffin143 commented:
@walragatver , it's working here too :)

birm commented:
I explicitly added @walragatver to this repository, so I believe that's it.

jeffin143 committed:
Merge https://github.com/mlpack/mlboard into cmake-setup

toshal-a commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,78 @@
+  void createEvent(size_t step, mlboard::Summary *summary);
+
+  /**
+   * A function to flush everything succefully and close the thread.
```
Typo error on this line. It should be `successfully`

toshal-a commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,78 @@
+  void writeSummary();
+
+  /**
+   * A helper function to chnage summary to event. The function should be
```
Typo error `change`

toshal-a commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,78 @@
+  std::string logDir() const { return logdir; }
+ private:
+  SharedQueue <mlboard::Event> q;
+  // std::thread does not have copy constructor hence pointer is safe
```
Missing, a period at the end of comment.

toshal-a commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,78 @@
+ * async operation which is running a separte thread. Look writeSummary()
+ * for more information.
+ */
+class fileWriter
```
I think we should use Pascal Case for class names and no case namespaces. Helps in a better understanding of code.

jeffin143 replied:
`no case namespaces` means ?

toshal-a replied:
Sorry, that means Lower Case namespace names.

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+
+fileWriter::fileWriter(std::string logdir,
+                       int maxQueueSize,
+                        std::size_t flushsec)
```
Misalignment.

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+
+void fileWriter::writeSummary()
+{
+  // this is a thread that will continously write summary one by one into file
```
A missing period at the end of the comment and yes the starting word should be capital.

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+  // this is a thread that will continously write summary one by one into file
+  while (true)
+  {
+    //  Break the loop if eveything is done
```
Missing period.

toshal-a commented on include/filewriter/sharedqueue.hpp:
```diff
@@ -0,0 +1,54 @@
+ * @tparam Datatype datatype of the elements queue would be holding. 
+ */
+template <typename Datatype>
+class SharedQueue
```
Over here you have used pascal case correctly. :)

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+
+void fileWriter::flush()
+{
+  // Flush everything succefully and close the thread.
```
Typo error `successfully`

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+      // wait for queue only if queue has any element or else it will
+      // forever
+      if (q.size() == 0) continue;
+      mlboard::Event event = q.pop();
```
I think we are only writing one object after exceeding flush sec right? I might be wrong, But I think we should write all objects present in the queue after crossing next time. Let me know what you think.

jeffin143 replied:
Yes , I am not sure what is the correct one pop element after flush sec or popping out everything between flush sec

Yes I think we should pop out multiple elements because it is wasting time, suppose flush sec is 4 sec and write took only 1 sec then it is wasting computer time for 3 sec wasting in the sense potential wasting is less but still 

I will change it

toshal-a replied:
https://github.com/tensorflow/tensorboard/blob/master/tensorboard/summary/writer/event_file_writer.py
I went through the Tensorflow's implementation it looks like they create a temp file in some file system and during flush they flush it out to the filesystem the entire content in the tempfile. 
In some file systems, they directly happened while writing itself so flushing would not have no effect.

jeffin143 replied:
I don't think so we need two different implementation. based on file system, Let me know how should we pan out the behaviour  here ?

I think so they did because of some limitation of python not sure

toshal-a replied:
Let's keep it simple. Flush will write all elements present at that time in the queue.

toshal-a commented on include/filewriter/summarywriter.hpp:
```diff
@@ -0,0 +1,43 @@
+{
+ public:
+  /**
+  * A function to create a scaler summary
```
A missing period at the end of the comment.

jeffin143 replied:
Done

toshal-a commented on include/filewriter/sharedqueue.hpp:
```diff
@@ -0,0 +1,54 @@
+  Datatype pop();
+
+  /**
+  * function to push an element in the queue.
```
I think the first letter of the comment should be capital

jeffin143 replied:
Done

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+    {
+      // wait for queue only if queue has any element or else it will
+      // forever
+      if (q.size() == 0) continue;
```
I think we are crosschecking it in q.pop() function right? Why are we validating once again over here? Have I missed something?

https://github.com/mlpack/mlboard/pull/1/files#diff-e5e0dbb8fc2e0677bbee3d2aefd78abcR15

jeffin143 replied:
if we remove the check this code will fail

```
int main()
{
    fileWriter f1("jeffin");
    f1.close();
}
```

reason is -> the event writer thread will try to pop an element but since there are 0 element it will wait 

And the notify will only work when a element is popped in ( If element is popped in then it will notify all the thread which are waiting, but since nothing is popped in it will be waiting infinitely , and hence this check

Another way of dealing this is closing  or destroying shared queue object which was intialized  in constructor, but I thought this would be safer rather than destroying since there could be elements in it :) 

Hope I cleared it, Let me know

toshal-a reviewed (commented):
Hi @jeffin143, 
I have left some comments. I am wondering whether we should also add a unit testing mechanism in the repo. I think we might need it in the future. I am not sure how much time consuming it is. So @jeffin143, @birm let me know your thoughts on it.

jeffin143 committed:
styling fix

jeffin143 committed:
use pascalcasing

jeffin143 committed:
change implementation of FileWriter::writeSummary()

jeffin143 committed:
specify to instal protobuf

birm commented on README.md:
```diff
@@ -1,2 +1,151 @@
+
+`brew install protobuf`
+
+- For Linux OS
```
Not to split hairs, but that's only for debian-like systems.

birm commented on include/filewriter/crc.hpp:
```diff
@@ -0,0 +1,16 @@
+/*
+**  CRC.H - header file for SNIPPETS CRC and checksum functions
```
Did you write this/these files?

jeffin143 replied:
/* Copyright (C) 1986 Gary S. Brown.  You may use this program, or
   code or tables extracted from it, as desired without restriction.*/

-> https://web.mit.edu/freebsd/head/sys/libkern/crc32.c

birm reviewed (approved):
Just some minor questions, but I think this is a reasonable start, and it certainly beats an empty repo :)

jeffin143 committed:
correct wording

toshal-a commented on README.md:
```diff
@@ -1,2 +1,151 @@
+    start = std::chrono::system_clock::now(); 
+    
+    // Creating a scaler summary
+    mlboard::SummaryWriter<mlboard::fileWriter>::scalar("tag",1,1.1,f1);
```
In this file, `fileWriter` should be replaced with `FileWriter`

jeffin143 replied:
Done

toshal-a commented on README.md:
````diff
@@ -1,2 +1,151 @@
+The above snippet of code would print the time taken for the scaler to log
+
+```
+elapsed time: 40.0163s
````
I think the elapsed time has changed now. I am getting 5.7157e-05 seconds. This is slightly weird, as default flushsec = 2.  Basically, It's because flushsec = 2 is in milliseconds.  After using `flushsec = 2000` I am getting `1.5 secs` on average as the elapsed time.

jeffin143 replied:
2 sec is too much time in my opinion and hence I choose 2 millisec, I you still want sec it as simple as using boost::second instead of std::chrono::seconds instead of std::chrono::milliseconds 

Let me know what you want

I agree that tensorflow has 2 sec , but I think cpp is faster than python and hence that time should also change and not follow same time

So let me know what you want

toshal-a replied:
Yes, you're correct according to language the time should go down. We need to ensure that it is not very much small. I think 2 milisec is very small. Basically we would be logging Metrics of ML models I think hardly any ML model would complete it's one step in 2 milisec.

Also, in tensorflow the default is 120000 milisec(2 mins). So I think we should reduce to 120000 / 10 = 12000 assuming c++ is 10x faster. I am fine with 2 secs all up to you.

https://www.tensorflow.org/api_docs/python/tf/summary/create_file_writer

jeffin143 replied:
Let's be in middle 5 sec :)

toshal-a commented on README.md:
````diff
@@ -1,2 +1,151 @@
+The total time of execution would be reduced to 10 sec since they all are executing parallelly.
+
+```
+elapsed time: 10.0004s
````
Same about this. I am now getting `elapsed time: 0.000584164s` as `flushsec = 2` milliseconds. After changing tof `flushSec = 2000`. I am getting the elapsed time on the same line which we are getting without using async. Let me know if I am going wrong.

jeffin143 replied:
I have changed the script a little bit, I was benchmarking it by editing the   function internally , I mean change the function implementation a little bit and then run the script

```
static void scalar(const std::string &tag,
                     int step,
                     double value,
                     filewriter &fw);
```


I agree that was confusing a little bit for the use so I made apt changed to the readme in last commit, Which will allow you to mirror the exact change

toshal-a commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,78 @@
+  *
+  * @param logdir Path to store the log files.
+  * @param maxQeueSize The maximum number of event to be store at a time.
+  * @param flushec Interval to perform the writing operation.
```
Typo error `flushsec`. Also flush sec is in seconds I guess. I think it's better to mention it in the comments.

jeffin143 replied:
done

toshal-a commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,78 @@
+  void close();
+
+  //! Get the path of log directory.
+  std::string logDir() const { return logdir; }
```
I think it's better to add a getter for getting `maxQueueSize` and `flushSec`. I think a setter would also be fine. Let me know what you think.

jeffin143 replied:
Done

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+      }
+      nexttime =
+          std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()
+          + std::chrono::milliseconds(flushsec));
```
If `flushsec` is in seconds. I think we need to multiply it by 1000. Let me know what you think.

jeffin143 replied:
I would like to keep it in miltisec, should I change it to `flushmilisec` ?

toshal-a replied:
Up to you. If we refer tensorflow, in tf1 `flushsec` parameter was in secs but in tf2 it's changed to `flushmilis` so it's up to you.

jeffin143 replied:
flushmilis

toshal-a commented on README.md:
```diff
@@ -1,2 +1,151 @@
+    std::chrono::duration<double> elapsed_seconds = end - start; 
+    std::time_t end_time = std::chrono::system_clock::to_time_t(end); 
+  
+    std::cout << "finished computation at " << std::ctime(&end_time) 
```
I think we should calculate the time after `f1.close()` as the thread finishes its execution after it. Let me know what you think.

jeffin143 replied:
I was just benchmarking the scalar creation function and not logging of the metrics 

We can benchmark that Once we are done with some more metric type like image so that we could benchmark it optimally

toshal-a replied:
Okay no problem.

toshal-a commented on README.md:
```diff
@@ -1,2 +1,151 @@
+    std::chrono::duration<double> elapsed_seconds = end - start; 
+    std::time_t end_time = std::chrono::system_clock::to_time_t(end); 
+  
+    std::cout << "finished computation at " << std::ctime(&end_time) 
```
Same about this I guess. I think we should calculate the final time after `f1.close()`

jeffin143 replied:
Currently it wouldn't make a much difference since , there are only 4 data points which is less than default maximum queue size

toshal-a replied:
Yes, you're correct. It's in readme at this moment. I was using the code as a test.  But I thought it's good to change it.
Actually when I was using before `f1.close()` and keeping `flushsec =2000` (2secs) I was getting time elapsed `0.0000587 ms`. And the program continues to  run. But it is wrong, as the thread is waiting to satisfy the condition.

jeffin143 replied:
Changed it accordingly

toshal-a reviewed (commented):
Hey, @jeffin143, the changes you have made look good. I have made some comments. Also, don't forget to rebase this PR as #3 is now merged.

jeffin143 committed:
Merge https://github.com/mlpack/mlboard into cmake-setup

jeffin143 committed:
Change wording of Readme

jeffin143 committed:
update readme and use flushmilis instead of flushsec

zoq commented on README.md:
```diff
@@ -1,2 +1,180 @@
+
+Following is a snippet that would log scaler for 4 steps in temp directory.
+
+You can compile the following snippet of code using : `g++ main.cpp -lproto -lprotobuf -lpthread` 
```
```suggestion
You can compile the following snippet using : `g++ main.cpp -lproto -lprotobuf -lpthread` 
```

Just a suggestion, to improve the flow.

jeffin143 replied:
Done

zoq commented on README.md:
````diff
@@ -1,2 +1,180 @@
+
+```
+#include <mlboard/mlboard.hpp>
+#include <iostream>
+#include <chrono> 
+#include <ctime> 
+#include <future>
````
If those header files are commonly used I guess it might be useful to include them directly in `mlboard/mlboard.hpp`?

jeffin143 replied:
#include <iostream> -> aldeardy there 
#include <chrono>  -> Only for benchmark
#include <ctime> --> Only for benchmark
#include <future> -> May be used often, but not sure

Let's wait for some more time and see what is often used :)

zoq commented on README.md:
```diff
@@ -1,2 +1,180 @@
+using namespace mlboard;
+
+// A function to mock that summary creation takes 10 sec.
+void mockfunc(const std::string &tag,
```
Do you mind to use the same style as for mlpack? `const std::string& tag` instead of `const std::string &tag`.

jeffin143 replied:
Done

zoq commented on README.md:
```diff
@@ -1,2 +1,180 @@
+int main()
+{
+    GOOGLE_PROTOBUF_VERIFY_VERSION;
+    // creating a FileWriter object that is repnsible for logging the summary.
```
```suggestion
    // creating a FileWriter object that is responsible for logging the summary.
```

jeffin143 replied:
Done

zoq commented on README.md:
```diff
@@ -1,2 +1,180 @@
+    std::chrono::time_point<std::chrono::system_clock> start, end; 
+    start = std::chrono::system_clock::now(); 
+    FileWriter f1("temp");
+    // Creating a scaler summary
```
```suggestion
    // Creating a scaler summary.
```

Pedantic style issue.

jeffin143 replied:
Done

zoq commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,78 @@
+  * @param flushmilis Interval to perform the writing operation (milliseconds).
+  */
+  FileWriter(std::string logdir,
+            int maxQueueSize = 10,
+            size_t flushmilis = 5000);
```
```suggestion
             int maxQueueSize = 10,
             size_t flushmilis = 5000);
```

jeffin143 replied:
Thanks :)

zoq commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,78 @@
+namespace mlboard {
+
+/**
+ * Class resposnible for writing the event to a event file. The writing is a
```
```suggestion
 * Class responsible for writing the event to a event file. The writing is a
```

jeffin143 replied:
Sorry :)

zoq commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,78 @@
+   * is intialised as a thread to complete the function without hindering
+   * any other option.
+   */
+  void writeSummary();
```
I would probably use the same style as for mlpack so `WriteSummary` instead of `writeSummary`.

jeffin143 replied:
Done

zoq commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,78 @@
+  void writeSummary();
+
+  /**
+   * A helper function to change summary to event. The function should be
```
```suggestion
   * A helper function to change summary to event. The function should
```

jeffin143 replied:
Done

zoq commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,78 @@
+
+  /**
+   * A helper function to change summary to event. The function should be
+   * always be there, since it is called from summary instance.
```
```suggestion
   * always be there, since it is called from the summary instance.
```

jeffin143 replied:
Done

jeffin143 committed:
update readme for grammer fixes and typos

jeffin143 committed:
add function to get and modify params and improve docs

toshal-a commented on include/filewriter/sharedqueue.hpp:
```diff
@@ -0,0 +1,54 @@
+  size_t MaxSize() const { return maxSize; }
+  //! Modify the maximum size of the queue.
+  size_t& MaxSize() { return maxSize; }
+  // ! Check if queue is empty
```
Missing period at the end

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+  while (true)
+  {
+    // Break the loop if eveything is done.
+    if (!(close_ || q.size() != 0)) break;
```
I think we should not break this loop when `q.size() == 0`. It might happen that queue becomes empty in the intermediate time. At that time future summary objects would not be logged.

toshal-a commented on include/filewriter/sharedqueue.hpp:
```diff
@@ -0,0 +1,54 @@
+  /**
+   * Function to push an element in the queue.
+   * 
+   * @param item The element to be pushed
```
Missing period at the end.

toshal-a commented on include/filewriter/sharedqueue_impl.hpp:
```diff
@@ -0,0 +1,42 @@
+void SharedQueue<Datatype>::push(const Datatype& item)
+{
+  std::unique_lock<std::mutex> mlock(mutex_);
+  while (queue_.size() > maxSize)
```
I think this should be >= maxSize.

toshal-a commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,86 @@
+  /**
+   * A function to flush everything successfully and close the thread.
+   */
+  void flush();
```
I am not sure of this. Should we also make `Flush()` instead of `flush()` like other functions.
```suggestion
  void Flush();
```

toshal-a commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,86 @@
+  /**
+   * Responsible for indicating the class to close the thread.
+   */ 
+  void close();
```
and same thing about over her `Close()`  function
```suggestion
  void Close();
```

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+    mockfunc("tag",2,2.1,f1);
+    mockfunc("tag",3,3.1,f1);
+    mockfunc("tag",4,4.1,f1);
+    f1.close();
```
Don't forget to change the readme syntax f1.
```suggestion
    f1.Close();
```

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+    result3.get();
+    result4.get();
+
+    f1.close();  
```
```suggestion
    f1.Close();  
```

toshal-a commented on include/filewriter/sharedqueue.hpp:
```diff
@@ -0,0 +1,54 @@
+  /**
+   * function to pop an element from the queue.
+   */ 
+  Datatype pop();
```
```suggestion
  Datatype Pop();
```

toshal-a commented on include/filewriter/sharedqueue.hpp:
```diff
@@ -0,0 +1,54 @@
+   * 
+   * @param item The element to be pushed
+   */
+  void push(const Datatype& item);
```
```suggestion
  void Push(const Datatype& item);
```

toshal-a commented on include/filewriter/sharedqueue.hpp:
```diff
@@ -0,0 +1,54 @@
+   */
+  void push(const Datatype& item);
+  //! Get the size of the queue.
+  size_t size() const { return queue_.size(); }
```
```suggestion
  size_t Size() const { return queue_.size(); }
```

toshal-a commented on include/filewriter/sharedqueue.hpp:
```diff
@@ -0,0 +1,54 @@
+  //! Modify the maximum size of the queue.
+  size_t& MaxSize() { return maxSize; }
+  // ! Check if queue is empty
+  bool empty() { return queue_.empty(); }
```
```suggestion
  bool Empty() { return queue_.empty(); }
```

toshal-a commented on include/filewriter/summarywriter.hpp:
```diff
@@ -0,0 +1,43 @@
+   * @param value Scaler value to be logged.
+   * @param fw filewriter object.
+   */
+  static void scalar(const std::string& tag,
```
```suggestion
  static void Scalar(const std::string& tag,
```

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+              FileWriter& fw)
+{
+    std::this_thread::sleep_for( std::chrono::seconds(10));
+    mlboard::SummaryWriter<mlboard::FileWriter>::scalar(tag,step,value,fw);
```
```suggestion
    mlboard::SummaryWriter<mlboard::FileWriter>::Scalar(tag,step,value,fw);
```

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+              FileWriter& fw)
+{
+    std::this_thread::sleep_for( std::chrono::seconds(10));
+    mlboard::SummaryWriter<mlboard::FileWriter>::scalar(tag,step,value,fw);
```
```suggestion
    mlboard::SummaryWriter<mlboard::FileWriter>::Scalar(tag,step,value,fw);
```

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+    q.push(event);
+}
+
+void FileWriter::flush()
```
```suggestion
void FileWriter::Flush()
```

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+  thread_->join();
+}
+
+void FileWriter::close()
```
```suggestion
void FileWriter::Close()
```

toshal-a commented on include/filewriter/sharedqueue_impl.hpp:
```diff
@@ -0,0 +1,42 @@
+namespace mlboard {
+
+template <typename Datatype>
+Datatype SharedQueue<Datatype>::pop()
```
```suggestion
Datatype SharedQueue<Datatype>::Pop()
```

toshal-a commented on include/filewriter/sharedqueue_impl.hpp:
```diff
@@ -0,0 +1,42 @@
+}
+
+template <typename Datatype>
+void SharedQueue<Datatype>::push(const Datatype& item)
```
```suggestion
void SharedQueue<Datatype>::Push(const Datatype& item)
```

toshal-a commented on include/filewriter/summarywriter_impl.hpp:
```diff
@@ -0,0 +1,28 @@
+namespace mlboard {
+
+template<typename filewriter>
+void SummaryWriter<filewriter>::scalar(const std::string& tag,
```
```suggestion
void SummaryWriter<Filewriter>::Scalar(const std::string& tag,
```

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+    {
+      while (q.size() > 0)
+      {
+        mlboard::Event event = q.pop();
```
```suggestion
        mlboard::Event event = q.Pop();
```

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+    if (timenow >= nexttime)
+    {
+      while (q.size() > 0)
```
```suggestion
      while (q.Size() > 0)
```

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+    event.set_wall_time(wall_time);
+    event.set_step(step);
+    event.set_allocated_summary(summary);
+    q.push(event);
```
```suggestion
    q.Push(event);
```

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+
+To generate summary data you can make use of `fileWriter` class and `SummaryWriter` class.
+
+For creating a summary, you have to call the specific summary type function from `SummaryWriter` class, for example to log a scaler summary you can call the scaler function as `mlboard::SummaryWriter<mlboard::fileWriter>::scalar(tag,step,value,filewriterobject);`. 
```
```suggestion
For creating a summary, you have to call the specific summary type function from `SummaryWriter` class, for example to log a scaler summary you can call the scaler function as: 
`mlboard::SummaryWriter<mlboard::FileWriter>::Scalar(tag,step,value,filewriterobject);` 
```

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+
+### 2. Usage
+
+To generate summary data you can make use of `fileWriter` class and `SummaryWriter` class.
```
```suggestion
To generate summary data you can make use of `FileWriter` class and `SummaryWriter` class.
```

toshal-a commented on include/filewriter/sharedqueue.hpp:
```diff
@@ -0,0 +1,54 @@
+{
+ public:
+  /**
+   * function to pop an element from the queue.
```
```suggestion
   * Function to pop an element from the queue.
```

toshal-a commented on include/filewriter/summarywriter.hpp:
```diff
@@ -0,0 +1,43 @@
+   * @param tag Tag to uniquely identify the scaler type.
+   * @param step The step at which scaler was logged.
+   * @param value Scaler value to be logged.
+   * @param fw filewriter object.
```
```suggestion
   * @param fw Filewriter object.
```

toshal-a commented on include/filewriter/summarywriter.hpp:
```diff
@@ -0,0 +1,43 @@
+ * @tparam filewriter The filewriter object whould would convert it
+ *    into events and then log to a file. 
+ */
+template<typename filewriter>
```
```suggestion
template<typename Filewriter>
```

toshal-a commented on include/filewriter/summarywriter.hpp:
```diff
@@ -0,0 +1,43 @@
+  static void scalar(const std::string& tag,
+                     int step,
+                     double value,
+                     filewriter& fw);
```
```suggestion
                     Filewriter& fw);
```

toshal-a commented on include/filewriter/summarywriter_impl.hpp:
```diff
@@ -0,0 +1,28 @@
+void SummaryWriter<filewriter>::scalar(const std::string& tag,
+                                       int step,
+                                       double value,
+                                       filewriter& fw)
```
```suggestion
                                       Filewriter& fw)
```

toshal-a commented on include/filewriter/summarywriter_impl.hpp:
```diff
@@ -0,0 +1,28 @@
+
+namespace mlboard {
+
+template<typename filewriter>
```
```suggestion
template<typename Filewriter>
```

toshal-a commented on include/filewriter/summarywriter_impl.hpp:
```diff
@@ -0,0 +1,28 @@
+/**
+ * @file filewriter.hpp
```
```suggestion
 * @file summarywriter_impl.hpp
```

toshal-a commented on include/filewriter/summarywriter.hpp:
```diff
@@ -0,0 +1,43 @@
+/**
+ * @file filewriter.hpp
```
```suggestion
 * @file summarywriter.hpp
```

toshal-a commented on include/filewriter/sharedqueue_impl.hpp:
```diff
@@ -0,0 +1,42 @@
+/**
+ * @file SharedQueue_impl.hpp
```
```suggestion
 * @file sharedqueue_impl.hpp
```

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+void FileWriter::close()
+{
+  close_ = false;
+  flush();
```
```suggestion
  Flush();
```

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+int main()
+{
+    GOOGLE_PROTOBUF_VERIFY_VERSION;
+    // creating a FileWriter object that is responsible for logging the summary.
```
```suggestion
    // Creating a FileWriter object that is responsible for logging the summary.
```

toshal-a commented on README.md:
````diff
@@ -1,2 +1,180 @@
+elapsed time: 44.6872s
+```
+
+Above timing is 40s (summary creation time) + 4s (flush timing i.e time taken to write event files)
````
```suggestion
Above timing is 40s (summary creation time) + 4s (flush timing i.e time taken to write event files).
```

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+
+You can then use tensorboard to visualize the scaler by using a simple command: `tensorboard --logdir .`
+
+The above snippet of code would print the time taken for the scaler to log
```
```suggestion
The above snippet of code would print the time taken for the scaler to log.
```

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+Make a build directory.  The directory can have any name, but 'build' is
+sufficient.
+
+    $ mkdir build
+    $ cd build
```
Add code block

```
$ mkdir build
$ cd build
```
instead of plain text.

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+The next step is to run CMake to configure the project.  Running CMake is the
+equivalent to running `./configure` with autotools. 
+
+    $ cmake ../
```
Add this in code block as well.

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+
+- For Mac OS 
+
+`brew install protobuf`
```
I think a code block better over here. you can insert a codeblock using ``` twice one before and after the code.
```
brew install protobuf
```

instead of 
`brew install protobuf`

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+
+- For debian-based systems
+
+`sudo apt-get install libprotobuf-dev protobuf-compiler`
```
Same about this. 
```
sudo apt-get install libprotobuf-dev protobuf-compiler
```
instead of 
`sudo apt-get install libprotobuf-dev protobuf-compiler`

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+
+mlboard is a tool that allows you to log mlpack data in the format that the TensorBoard frontend can render in browsers.
+
+Mlboard is still in the development phase. You can track the development ideas [here](https://www.mlpack.org/gsocblog/Jeffin2020CBP.html)
```
```suggestion
mlboard is still in the development phase. You can track the development ideas [here](https://www.mlpack.org/gsocblog/Jeffin2020CBP.html)
```

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+
+Once CMake is configured, building the library is as simple as typing 'make'.
+
+    $ make
```
Add in code block 
```
$ make 
```

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+and `/usr/local/bin/`, make sure you have root privileges (or write permissions 
+to those three directories), and simply type
+
+    $ make install
```
Add code block
```
$ make install
```

toshal-a replied:
Looks like you missed this.

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+
+Above timing is 40s (summary creation time) + 4s (flush timing i.e time taken to write event files)
+
+Alternatively you can execute the logging in an async manner which would be faster at many instances by using `std::async`, here is a snippet which allows you to do the same operation but much faster as compared to the previous code
```
```suggestion
Alternatively you can execute the logging in an async manner which would be faster at many instances by using `std::async`, here is a snippet which allows you to do the same operation but much faster as compared to the previous code.
```

toshal-a commented on README.md:
```diff
@@ -1,2 +1,180 @@
+
+As you can see from both the snippet, the flush timing remains same since deafult flush timing are `5000 milliseconds` and we are able to get approx same flush time, for both of the above code. The significant reduction was in summary creation time becuase of async function calling.
+
+Note: Just to benchmark, a waiting time of 10 sec was added using `std::this_thread::sleep_for( std::chrono::seconds(10));` inside the `mockfunc` (to mock a behavior of writing a summary which has a lot of data), so that there could be a clear difference between the two codes 
```
```suggestion
Note: Just to benchmark, a waiting time of 10 sec was added using `std::this_thread::sleep_for( std::chrono::seconds(10));` inside the `mockfunc` (to mock a behavior of writing a summary which has a lot of data), so that there could be a clear difference between the two codes.
```

toshal-a reviewed (commented):
Hi @jeffin143,
I think this would be probably the last review. The code looks logically correct :+1: . There are some style issues which needs to be addressed.

toshal-a commented on include/filewriter/filewriter_impl.hpp:
```diff
@@ -0,0 +1,91 @@
+        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+    if (timenow >= nexttime)
+    {
+      while (q.size() > 0)
```
Okay, this has just clicked my mind. Basically we are using a mutex to gain control over the queue and to synchronize its access. We are acquiring and releasing the lock multiple times while emptying (in the `ShareQueue::Pop()`). Because of this, two things can happen

1.  it might happen that someone (`the main thread`) pushes into the queue by acquiring lock in the middle of clearing the queue. This might lead to some other issues.
2. It is also not optimum to acquire and release the mutex lock multiple times. 

So, I think it's better that we create a function in `SharedQueue` itself to empty it. A function like `ClearQueue()/ FlushQueue()` would acquire the lock once in the function call and will release it once the queue becomes empty and all objects are written to the file. 

 Let me know what you think about it.

jeffin143 replied:
Then it will lock the queue for a longer instance and we won't be able to push it to the queue  delaying the training if we are pushing in our callbacks between the training instance

toshal-a replied:
Yeah, you are correct. I forgot about the call backs. Let's keep it as it is.

toshal-a reviewed (commented):
Hey, I was just wondering about the code and I found something. Let me know your thoughts on it.

jeffin143 committed:
Fix styling and apply suggestions

jeffin143 commented:
@walragatver I have resolved the comments but I don't know there is a small issue 

If you use the code suggested in `readme.md` and then remove f1.Close() from it the thread still terminates 

I am not sure why that happens because I have given it `while(true)`, that means it should run infinitely many time

Try it with plotting 100 data points and still it exists after writing all those 100 points

jeffin143 commented:
> If you use the code suggested in `readme.md` and then remove f1.Close() from it the thread still terminates
> 
 
Found out the reason , if you don't use f1.Close(), it will abruptly close the the thread since it goes out of scope and hence The solution is we can declare a destructor which will close the function when it goes out of scope

jeffin143 committed:
flush everything which is pending

jeffin143 commented:
> I think we should not break this loop when q.size() == 0. It might happen that queue becomes empty in the intermediate time. At that time future summary objects would not be logged.

The loop will only break if you `close_ = false` , and you can only make it false by calling `f1.Close()` function 

Now if we remove q.size()==0 condition and if user calls f1.Close() function then it will close the loop then and there, whereas it should not behave like that it should continue till the queue is empty and when queue is empty then it should break, A user should not call f1.close() function if he want's to log more, he should only call it when everything has been logged 

Please try removing the condition and try to run the above code it will only log 2 scaler values instead of 4 because , the close_ will make the loop terminate at that instance but there will be 2 events in the queue waiting to be written and hence I added that condition , Refer the problem in previous comment


Now I have added a destructor also, which will allow you to log properly even if filewriter object goes out of object, Then it will only stop writing if the queue is empty

toshal-a commented:
> > I think we should not break this loop when q.size() == 0. It might happen that queue becomes empty in the intermediate time. At that time future summary objects would not be logged.
> 
> The loop will only break if you `close_ = false` , and you can only make it false by calling `f1.Close()` function
> 
> Now if we remove q.size()==0 condition and if user calls f1.Close() function then it will close the loop then and there, whereas it should not behave like that it should continue till the queue is empty and when queue is empty then it should break, A user should not call f1.close() function if he want's to log more, he should only call it when everything has been logged
> 
> Please try removing the condition and try to run the above code it will only log 2 scaler values instead of 4 because , the close_ will make the loop terminate at that instance but there will be 2 events in the queue waiting to be written and hence I added that condition , Refer the problem in previous comment
> 
> Now I have added a destructor also, which will allow you to log properly even if filewriter object goes out of object, Then it will only stop writing if the queue is empty

Yeah I misinterpreted it earlier :+1:

toshal-a commented on README.md:
```diff
@@ -1,2 +1,191 @@
+To generate summary data you can make use of `FileWriter` class and `SummaryWriter` class.
+
+For creating a summary, you have to call the specific summary type function from `SummaryWriter` class, for example to log a scaler summary you can call the scaler function as:
+`mlboard::SummaryWriter<mlboard::fileWriter>::scalar(tag,step,value,filewriterobject);`. 
```
```suggestion
`mlboard::SummaryWriter<mlboard::FileWriter>::Scalar(tag,step,value,filewriterobject);`. 
```

jeffin143 replied:
Done

toshal-a commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,91 @@
+  //! Modify the maximum size of the queue.
+  size_t& MaxSize() { return q.MaxSize(); }
+ private:
+  SharedQueue<mlboard::Event> q;
```
Can you add comment on every data member like the way added below? A simple statement would be sufficient.
https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/ffn.hpp#L459

toshal-a commented on include/filewriter/sharedqueue.hpp:
```diff
@@ -0,0 +1,54 @@
+  // ! Check if queue is empty.
+  bool Empty() { return queue_.empty(); }
+ private:
+  std::queue<Datatype> queue_;
```
Same over here I think comments would lead to a better understanding to a novice.

jeffin143 replied:
Done

toshal-a commented on include/filewriter/sharedqueue.hpp:
```diff
@@ -0,0 +1,54 @@
+  size_t MaxSize() const { return maxSize; }
+  //! Modify the maximum size of the queue.
+  size_t& MaxSize() { return maxSize; }
+  // ! Check if queue is empty.
```
```suggestion
  //! Check if queue is empty.
```

toshal-a commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,91 @@
+
+  //! Get the path of log directory.
+  std::string LogDir() const { return logdir; }
+  //! Modify the flusmilis.
```
```suggestion
  //! Modify the flushmilis.
```

toshal-a commented on include/filewriter/filewriter.hpp:
```diff
@@ -0,0 +1,91 @@
+  std::string LogDir() const { return logdir; }
+  //! Modify the flusmilis.
+  size_t& FlushMilis() { return flushmilis; }
+  //! Get the flusmilis.
```
```suggestion
  //! Get the flushmilis.
```

toshal-a commented on README.md:
````diff
@@ -1,2 +1,191 @@
+elapsed time: 14.0004s
+```
+
+Above timing is 10s (summary creation time) + 4s (flush timing i.e time taken to write event files )
````
```suggestion
Above timing is 10s (summary creation time) + 4s (flush timing i.e time taken to write event files).
```

toshal-a commented on README.md:
```diff
@@ -1,2 +1,191 @@
+
+As you can see from both the snippet, the flush timing remains same since deafult flush timing are `5000 milliseconds` and we are able to get approx same flush time, for both of the above code. The significant reduction was in summary creation time becuase of async function calling.
+
+Note: Just to benchmark, a waiting time of 10 sec was added using `std::this_thread::sleep_for( std::chrono::seconds(10));` inside the `mockfunc` (to mock a behavior of writing a summary which has a lot of data), so that there could be a clear difference between the two codes .
```
```suggestion
Note: Just to benchmark, a waiting time of 10 sec was added using `std::this_thread::sleep_for( std::chrono::seconds(10));` inside the `mockfunc` (to mock a behavior of writing a summary which has a lot of data), so that there could be a clear difference between the two codes.
```

toshal-a reviewed (approved):
Hi, this looks good. :+1: These are my final comments. This is ready to be merged.

jeffin143 committed:
update readme and add docs
